src/Microsoft.SqlTools.ServiceLayer/Formatter/TSqlFormatterService.cs

```diff
@@ -0,0 +1,270 @@
+        {
+            return await Task.Factory.StartNew(() =>
+            {
+                var range = docFormatParams.Range;
+                ScriptFile scriptFile = GetFile(docFormatParams);
+                TextEdit textEdit = new TextEdit { Range = range };
+                string text = scriptFile.GetTextInRange(range.ToBufferRange());
+                return DoFormat(docFormatParams, textEdit, text);
+            });
+        }
+
+        private async Task<TextEdit[]> FormatAndReturnEdits(DocumentFormattingParams docFormatParams)
+        {
+            return await Task.Factory.StartNew(() =>
+            {
+                var scriptFile = GetFile(docFormatParams);
+                TextEdit textEdit = PrepareEdit(scriptFile);
+                string text = scriptFile.Contents;
+                return DoFormat(docFormatParams, textEdit, text);
+            });
+        }
+
+        private TextEdit[] DoFormat(DocumentFormattingParams docFormatParams, TextEdit edit, string text)
+        {
+            Validate.IsNotNull(nameof(docFormatParams), docFormatParams);
+
+            FormatOptions options = GetOptions(docFormatParams);
+            List<TextEdit> edits = new List<TextEdit>();
+            edit.NewText = Format(text, options, false);
+            // TODO do not add if no formatting needed?
+            edits.Add(edit);
+            return edits.ToArray();
+        }
+
+        private FormatOptions GetOptions(DocumentFormattingParams docFormatParams)
+        {
+            FormatOptions options = new FormatOptions();
+            if (docFormatParams.Options != null)
+            {
+                options.UseSpaces = docFormatParams.Options.InsertSpaces;
+                options.SpacesPerIndent = docFormatParams.Options.TabSize;
+            }
+            if (settings != null)
+            {
+                if (settings.AlignColumnDefinitionsInColumns.HasValue) { options.AlignColumnDefinitionsInColumns = settings.AlignColumnDefinitionsInColumns.Value; }
+
+                if (settings.PlaceCommasBeforeNextStatement.HasValue) { options.PlaceCommasBeforeNextStatement = settings.PlaceCommasBeforeNextStatement.Value; }
+
+                if (settings.PlaceSelectStatementReferencesOnNewLine.HasValue) { options.PlaceEachReferenceOnNewLineInQueryStatements = settings.PlaceSelectStatementReferencesOnNewLine.Value; }
+
+                if (settings.UseBracketForIdentifiers.HasValue) { options.EncloseIdentifiersInSquareBrackets = settings.UseBracketForIdentifiers.Value; }
+
+                options.DatatypeCasing = settings.DatatypeCasing;
+                options.KeywordCasing = settings.KeywordCasing;
+            }
+            return options;
+        }
+        
+        private ScriptFile GetFile(DocumentFormattingParams docFormatParams)
+        {
+            if (WorkspaceService == null)
+            {
+                // TODO throw
```

Comment:
>// TODO throw [](start = 16, length = 13)

why not just throw since the next line is going to throw anyways...or delete the if and just let the following line throw